### PR TITLE
Mandd/ensemble functionsfixes

### LINKED ISSUE #198: Better error message when extraneous inputs are passed as <Input> of the Code interface
--------
Issue Description
--------
##### What did you expect to see happen?
We received the following question on the user list:
----------------------------------------------
Hi RAVEN users,

I am having troubles to have RAVEN MultipleRuning RELAP5-3D.

The RELAP5 input file I am unsuccesfully run is "ans79.i: one of the RELAP5 input files provided with the installation CD and used also to certify the installation.

Using the following input file, I would like to:
Vary both the total power and the initial pressure inside the pipe,
Run 10 RELAP5 cases,
Plot some results - both pointset and historyset types

The actual result is an error message which is not really clear, could you please help me understand what I am doing wrong?

<?xml version="1.0" ?>
<Simulation verbosity="debug">

  <RunInfo>
    <JobName>ForwardSampl_ans79</JobName>
    <WorkingDir>.</WorkingDir>
    <batchSize>4</batchSize>
    
    <Sequence>GenerateData,writePointSet,writehistory</Sequence>
  </RunInfo>

  <TestInfo>
    <name>RavenMultipleRunRelap5</name>
    <author>av</author>
    <created>2017-05-22</created>
  </TestInfo>

  <Files>
    <Input name="inputFile"   type="" >ans79.i</Input>
    <Input name="tpfh2o"      type="" >tpfh2o</Input>
    <Input name="license"     type="" >rellic.bin</Input>
  </Files>

  <Models>
    <Code name="r5model" subType="Relap5">
      <executable>./relap5.x</executable>
      <clargs type="text" arg=" -r restrt.r"/>
    </Code>
  </Models>

  <Steps>
    <MultiRun name="GenerateData">
      <Sampler class="Samplers"    type="MonteCarlo"   >myMC</Sampler>

      <Input   class="DataObjects" type="PointSet"     >DummyIN</Input>
      
      <Input    class="Files"   type=""     >inputFile</Input>
      <Input    class="Files"   type=""     >tpfh2o</Input>
      <Input    class="Files"   type=""     >license</Input>
      
      <Model    class="Models"  type="Code" >r5model</Model>
      
      <Output   class="DataObjects"  type="PointSet"  >pointValues</Output>
      <Output   class="DataObjects"  type="HistorySet">history</Output>
    </MultiRun>
    
    <IOStep name="writePointSet" pauseAtEnd="True">
        <Input class="DataObjects" type="PointSet">pointValues</Input>
        <Output class="OutStreams" type="Print">history</Output>
        <Output class="OutStreams" type="Plot">historyPlot</Output>
    </IOStep>
    
    <IOStep name="writehistory" pauseAtEnd="True">
        <Input class="DataObjects" type="HistorySet">history</Input>
        <Output class="OutStreams" type="Print">history</Output>
        <Output class="OutStreams" type="Plot">historyPlot</Output>
    </IOStep>
  </Steps>

  <Distributions>
    <Normal name="Normal">
      <mean>1.0e+6</mean>
      <sigma>2.5e+4</sigma>
      <lowerBound>9.95e+5</lowerBound>
      <upperBound>1.05e+6</upperBound>
    </Normal>
    
    <Uniform name="Uniform">
      <lowerBound>7.0e+6</lowerBound>
      <upperBound>7.5e+6</upperBound>
    </Uniform>
  </Distributions>

  <Samplers>
    <MonteCarlo name="myMC">
      <samplerInit>
        <limit>10</limit>
      </samplerInit>
      
      <variable name="30000001:2">
        <distribution>Normal</distribution>
      </variable>
      <variable name="0031201:2">
        <distribution>Uniform</distribution>
      </variable>
      
    </MonteCarlo>
  </Samplers>

  <DataObjects>
    <PointSet name="DummyIN">
      <Input>30000001:2,0031201:2</Input>
    </PointSet>

    <PointSet name="pointValues">
      <Input>30000001:2,0031201:2</Input>
      <Output>rktpow,rkreaqc</Output>
    <options>
      <outputRow>-3</outputRow>
    </options>
    </PointSet>
    
    <HistorySet name="history">
      <Input>30000001:2,0031201:2</Input>
      <Output>time,rkfipow_0,rkgapow_0,rkreac_0</Output>
    </HistorySet>
  </DataObjects>

  <OutStreams>
    <Print name="pointValues">
      <type>csv</type>
      <source>pointValues</source>
    </Print>
    
    <Print name="history">
      <type>csv</type>
      <source>history</source>
    </Print>
    
    <Plot   name="historyPlot" verbosity="debug">
      <plotSettings>
        <plot>
          <type>line</type>
          <x>history|Output|time</x>
          <y>history|Output|rkfipow_0</y>
          <c>blue</c>
        </plot>
        <plot>
          <type>line</type>
          <x>history|Output|time</x>
          <y>history|Output|rkgapow_0</y>
          <c>red</c>
        </plot>
        <plot>
          <type>line</type>
          <x>history|Output|time</x>
          <y>history|Output|rkreac_0</y>
          <c>yellow</c>
        </plot>
        <xlabel>time (s)</xlabel>
        <ylabel>power (W)</ylabel>
      </plotSettings>

      <actions>
        <how>png,screen</how>
        <title>
          <text> </text>
        </title>
      </actions>
    </Plot>
    
    <Plot  name="myPlot">
      <plotSettings>
        <plot>
          <type>scatter</type>
          <x>pointValues|Input|30000001:2</x>
          <y>pointValues|Input|0031201:2</y>
          <z>pointValues|Output|rktpow_0</z>
          <colorMap>pointValues|Output|rkreaqc_0</colorMap>
        </plot>
        <xlabel>Total Power - input (W)</xlabel>
        <ylabel>Pipe Initial Pressure (psi)</ylabel>
        <zlabel>Total Power - output (W)</zlabel>
      </plotSettings>
      <actions><how>png,screen</how></actions>
    </Plot>
  
  </OutStreams>

</Simulation>

Error message:
(    0.03 sec) STEP MULTIRUN            : DEBUG           -> jobHandler initialized
Traceback (most recent call last):
  File "/home/av_ale/raven/framework/Driver.py", line 278, in <module>
    raven()
  File "/home/av_ale/raven/framework/Driver.py", line 231, in raven
    simulation.run()
  File "/home/av_ale/raven/framework/Simulation.py", line 1051, in run
    stepInstance.takeAstep(stepInputDict)
  File "/home/av_ale/raven/framework/Steps.py", line 235, in takeAstep
    self._initializeStep(inDictionary)
  File "/home/av_ale/raven/framework/Steps.py", line 191, in _initializeStep
    self._localInitializeStep(inDictionary)
  File "/home/av_ale/raven/framework/Steps.py", line 462, in _localInitializeStep
    SingleRun._localInitializeStep(self,inDictionary)
  File "/home/av_ale/raven/framework/Steps.py", line 335, in _localInitializeStep
    inDictionary['Model'].initialize(inDictionary['jobHandler'].runInfoDict,inDictionary['Input'],modelInitDict)
  File "/home/av_ale/raven/framework/Models.py", line 1440, in initialize
    shutil.copy(inputFile.getAbsFile(),self.workingDir)
AttributeError: 'PointSet' object has no attribute 'getAbsFile'
------------------------------------------------------------------------------

##### What did you see instead?

The code should error out in a more understandable way when extraneous inputs are passed as Input of the Step.

##### Do you have a suggested fix for the development team?

Check

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [X] 1. Is it tagged with a type: defect or improvement? ***improvement***
- [X] 2. Is it tagged with a priority: critical, normal or minor? ***normal***
- [X] 3. If it will impact requirements or requirements tests, is it tagged with requirements? ***n/a***
- [X] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users. ***n/a***
- [X] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.) ***yes***

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed? ***improvement***
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)   ***improvement***
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?  ***improvement***
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?  ***improvement***
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?  ***PR #205***

### LINKED ISSUE #206: Functions for ensemble sampler
--------
Issue Description
--------
##### What did you expect to see happen?
variables determined from functions should be computed

##### What did you see instead?
They are not computed

##### Please attach the input file(s) that generate this error.  The simpler the input, the faster we can find the issue.

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement? defect
- [x] 2. Is it tagged with a priority: critical, normal or minor? normal
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements? no
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users. No
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.) yes

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed? yes
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.) Yes, test has been added
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)? NA
- [x] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)? NA
- [x] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided? NA

### PR DESCRIPTION
**Pull request by mandd**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #206 
Closes #198
##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

File: framework/Samplers/EnsembleForward.py [hunk 1]
```diff
@@ -68,10 +69,16 @@ def localInputAndChecks(self,xmlNode):
...
-        self.raiseAnError(IOError,"Sampling strategy "+child.tag+" is not usable in "+self.type+" Sampler. Available are "+",".join(self.acceptableSamplers))
-      else:
-        self.raiseAnError(IOError,"XML node "+ child.tag + " unknown. Check the Manual!")
+      elif child.tag=='variable':
+        for childChild in child:
+          if childChild.tag == 'function':
+            self.dependentSample[child.attrib['name']] = childChild.text
+          else:
+            self.raiseAnError(IOError,"Error")
+      #elif child.tag in knownTypes():
```
Comment (alfoa) [inline framework/Samplers/EnsembleForward.py]: can you remove the commented lines?

File: framework/Samplers/EnsembleForward.py [hunk 2]
```diff
@@ -68,10 +69,16 @@ def localInputAndChecks(self,xmlNode):
         self.instanciatedSamplers[child.tag] = returnInstance(child.tag,self)
         #FIXME the variableGroups needs to be fixed
         self.instanciatedSamplers[child.tag].readXML(child,self.messageHandler,variableGroups={},globalAttributes=self.globalAttributes)
-      elif child.tag in knownTypes():
```
Comment (alfoa) [inline framework/Samplers/EnsembleForward.py]: what about the Sampler we specify in the EnsembleForward is not among the known one?

File: framework/Samplers/EnsembleForward.py [hunk 3]
```diff
@@ -68,10 +69,16 @@ def localInputAndChecks(self,xmlNode):
...
-        self.raiseAnError(IOError,"Sampling strategy "+child.tag+" is not usable in "+self.type+" Sampler. Available are "+",".join(self.acceptableSamplers))
-      else:
-        self.raiseAnError(IOError,"XML node "+ child.tag + " unknown. Check the Manual!")
+      elif child.tag=='variable':
+        for childChild in child:
+          if childChild.tag == 'function':
+            self.dependentSample[child.attrib['name']] = childChild.text
+          else:
+            self.raiseAnError(IOError,"Error")
```
Comment (alfoa) [inline framework/Samplers/EnsembleForward.py]: Error message should be more specific

File: framework/Samplers/EnsembleForward.py [hunk 4]
```diff
@@ -104,6 +111,15 @@ def _localGenerateAssembler(self,initDict):
       combSampler._localGenerateAssembler(initDict)
     self.raiseADebug("Distributions initialized!")
 
+    for key,val in self.dependentSample.items():
+      if val not in availableFunc.keys():
+        self.raiseAnError('Function',val,'was not found among the available functions:',availableFunc.keys())
```
Comment (alfoa) [inline framework/Samplers/EnsembleForward.py]: which exception ? ```IOError```?

Comment (alfoa) [inline framework/Samplers/EnsembleForward.py]: I think this call is not well-formed

File: framework/Samplers/EnsembleForward.py [hunk 5]
```diff
@@ -162,4 +180,17 @@ def localGenerateInput(self,model,myInput):
...
+    self.inputInfo['SamplerType'] = 'EnsembleForward' 
+    
+    # Update dependent variables  
+    for var in self.dependentSample.keys():
+      test=self.funcDict[var].evaluate("evaluate",self.inputInfo['SampledVars'])
+      for corrVar in var.split(","):
+        self.values[corrVar.strip()] = test
+        self.inputInfo['SampledVars'][corrVar.strip()] = test
+    
+    #print(self.inputInfo['SampledVars'])
```
Comment (alfoa) [inline framework/Samplers/EnsembleForward.py]: commented line

File: framework/Samplers/EnsembleForward.py (hunk 2 above)
Comment (mandd) [inline framework/Samplers/EnsembleForward.py]: Edit: i left it out during the development.
It is now back

File: framework/Samplers/EnsembleForward.py (hunk 3 above)
Comment (mandd) [inline framework/Samplers/EnsembleForward.py]: Edited

File: framework/Samplers/EnsembleForward.py (hunk 1 above)
Comment (mandd) [inline framework/Samplers/EnsembleForward.py]: removed #

File: framework/Samplers/EnsembleForward.py (hunk 4 above)
Comment (mandd) [inline framework/Samplers/EnsembleForward.py]: edited

**Comment by alfoa** [comment]
----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code. ***ok***
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts). ***ok***
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details. ***ok***
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass. ***they do***
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True. ***test added***
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync. ***n/a***
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done. ***#206 #198***
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added? ***n/a***

**Comment by alfoa** [comment]
Check list passed...

PR can be accepted...

**Pull request merged**